File: typescript/src/ancilis/producers/crewai.ts
/**
 * CrewAI framework producer (TypeScript parity with `ancilis.producers.crewai`).
 *
 * CrewAI's official surface is Python-only, but TS parity preserves the
 * ergonomics for users who run mixed-language stacks or wrap CrewAI through
 * a polyglot proxy. The producer exposes step / task / crew callback factories
 * with the same signatures the Python SDK uses.
 */

import { createHash, randomUUID } from "node:crypto";
import type { ResolvedConfig } from "../config/index.js";
import type { Action } from "../engine/action.js";
import { Engine } from "../engine/engine.js";
import { ToolRegistry, ToolStatus } from "../engine/registry.js";
import type { ToolEntry } from "../engine/registry.js";
import type { EvaluationResult } from "../engine/result.js";
import { matchesToolList } from "../engine/tool-matching.js";
import { EvidenceStore } from "../evidence/store.js";
import { canonicalJsonStringify } from "../evidence/chain.js";
import { recordAdapterUsed } from "../telemetry/index.js";
import { ProducerType } from "./protocol.js";

const PROVIDER = "crewai";
const PRODUCER_VERSION = "0.1.0";

export type CrewAIEventKind = "step" | "task" | "crew";

export interface CrewAIEvent {
  kind: CrewAIEventKind;
  name: string;
  agentName: string;
  output?: unknown;
  metadata?: Record<string, unknown>;
}

export interface CrewAIObservation {
  action: Action;
  evaluation: EvaluationResult;
}

function firstStringAttr(obj: unknown, attrs: readonly string[]): string | null {
  if (obj === null || obj === undefined || typeof obj !== "object") return null;
  const o = obj as Record<string, unknown>;
  for (const attr of attrs) {
    const v = o[attr];
    if (typeof v === "string" && v) return v;
  }
  return null;
}

function firstStringKey(obj: unknown, keys: readonly string[]): string | null {
  if (obj === null || obj === undefined || typeof obj !== "object" || Array.isArray(obj)) {
    return null;
  }
  const o = obj as Record<string, unknown>;
  for (const k of keys) {
    const v = o[k];
    if (typeof v === "string" && v) return v;
  }
  return null;
}

const STEP_KEYS = ["tool", "tool_name", "agent_role", "agent", "name"] as const;
const TASK_KEYS = ["description", "task_id", "name", "agent_role"] as const;
const CREW_KEYS = ["name", "id", "crew_id"] as const;

export function _stepName(stepOutput: unknown, fallback: string): string {
  return (
    firstStringAttr(stepOutput, STEP_KEYS) ??
    firstStringKey(stepOutput, STEP_KEYS) ??
    fallback
  );
}

export function _taskName(taskOutput: unknown, fallback: string): string {
  return (
    firstStringAttr(taskOutput, TASK_KEYS) ??
    firstStringKey(taskOutput, TASK_KEYS) ??
    fallback
  );
}

export function _crewName(crewOutput: unknown, fallback: string): string {
  return (
    firstStringAttr(crewOutput, CREW_KEYS) ??
    firstStringKey(crewOutput, CREW_KEYS) ??
    fallback
  );
}

export function _serializable(value: unknown): unknown {
  if (
    value === null ||
    value === undefined ||
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean" ||
    Array.isArray(value)
  ) {
    return value;
  }
  if (typeof value === "object") {
    const o = value as Record<string, unknown>;
    for (const method of ["modelDump", "model_dump", "dict", "toDict", "to_dict"]) {
      const fn = o[method];
      if (typeof fn === "function") {
        try {
          return (fn as () => unknown).call(value);
        } catch {
          continue;
        }
      }
    }
    // Plain object — return as-is for JSON serialization.
    return value;
  }
  return String(value);
}

export class CrewAIActionProducer {
  protected _config: ResolvedConfig;
  protected _engine: Engine;
  protected _registry: ToolRegistry;
  protected _evidenceStore: EvidenceStore;
  private _sessionId: string = randomUUID();

  constructor(
    config: ResolvedConfig,
    engine: Engine,
    registry?: ToolRegistry,
    evidenceStore?: EvidenceStore,
  ) {
    this._config = config;
    this._engine = engine;
    this._registry = registry ?? engine.registry;
    this._evidenceStore = evidenceStore ?? new EvidenceStore(config);
    recordAdapterUsed(PROVIDER);
  }

  get config(): ResolvedConfig { return this._config; }
  get producerType(): ProducerType { return ProducerType.FRAMEWORK; }
  get producerVersion(): string { return PRODUCER_VERSION; }
  get sessionId(): string { return this._sessionId; }

  protected _toolName(event: CrewAIEvent): string {
    return `${PROVIDER}:${event.kind}:${event.name}`;
  }

  protected _buildDcCodes(): string[] {
    const codes: string[] = [];
    for (const dcCodes of this._config.dataClassifications.values()) {
      for (const code of dcCodes) {
        if (!codes.includes(code)) codes.push(code);
      }
    }
    return codes;
  }

  translate(event: CrewAIEvent): Action {
    const payload: Record<string, unknown> = {
      provider: PROVIDER,
      kind: event.kind,
      name: event.name,
      output: _serializable(event.output),
      metadata: event.metadata ?? {},
    };
    const paramHash = createHash("sha256")
      .update(canonicalJsonStringify(payload))
      .digest("hex");
    const toolName = this._toolName(event);
    const entry = this._registry.lookup(toolName);
    return {
      actionId: randomUUID(),
      timestamp: new Date().toISOString(),
      agentId: event.agentName,
      sourceType: this.producerType,
      producerType: this.producerType,
      producerVersion: this.producerVersion,
      agentOwner: this._config.agentOwner ?? null,
      actionType: event.kind === "step" ? "tool_call" : "api_request",
      tool: {
        name: toolName,
        server: PROVIDER,
        descriptionHash: entry?.descriptionHash ?? null,
      },
      parameters: { raw: payload, parameterHash: paramHash },
      context: {
        dataClassifications: this._buildDcCodes(),
        activeOverlays: [...this._config.activeOverlays.keys()],
      },
    };
  }

  computeToolHash(toolIdentifier: string): string {
    return createHash("sha256").update(toolIdentifier).digest("hex");
  }

  registerTools(registry: ToolRegistry): string[] {
    return registry.getAll().map((entry) => entry.name);
  }

  protected _ensureRegistered(event: CrewAIEvent): string {
    const name = this._toolName(event);
    if (!this._registry.lookup(name)) {
      const status = matchesToolList(name, this._config.toolsAllowed)
        ? ToolStatus.APPROVED
        : ToolStatus.OBSERVED;
      const now = new Date().toISOString();
      this._registry.register({
        name,
        descriptionHash: this.computeToolHash(name),
        status,
        approvedBy: status === ToolStatus.APPROVED ? "config" : null,
        firstSeen: now,
        statusChanged: now,
      } satisfies ToolEntry);
    }
    return name;
  }

  async observe(event: CrewAIEvent): Promise<CrewAIObservation> {
    const toolName = this._ensureRegistered(event);
    const action = this.translate(event);
    const evaluation = this._engine.evaluate(action);
    await this._evidenceStore.store(evaluation, toolName);
    return { action, evaluation };
  }

  // --- CrewAI-shaped callback factories ---

  /** Returns a callback matching CrewAI's `step_callback` signature. */
  stepCallback(agentName?: string): (stepOutput: unknown) => Promise<void> {
    const agent = agentName ?? this._config.agentName;
    return async (stepOutput: unknown): Promise<void> => {
      await this.observe({
        kind: "step",
        name: _stepName(stepOutput, "step"),
        agentName: agent,
        output: stepOutput,
      });
    };
  }

  /** Returns a callback for the per-Task `callback=` argument. */
  taskCallback(agentName?: string): (taskOutput: unknown) => Promise<void> {
    const agent = agentName ?? this._config.agentName;
    return async (taskOutput: unknown): Promise<void> => {
      await this.observe({
        kind: "task",
        name: _taskName(taskOutput, "task"),
        agentName: agent,
        output: taskOutput,
      });
    };
  }

  /** Returns a callback suitable for crew-level step/completion callbacks. */
  crewCallback(agentName?: string): (crewOutput: unknown) => Promise<void> {
    const agent = agentName ?? this._config.agentName;
    return async (crewOutput: unknown): Promise<void> => {
      await this.observe({
        kind: "crew",
        name: _crewName(crewOutput, "crew"),
        agentName: agent,
        output: crewOutput,
      });
    };
  }
}

File: python/src/ancilis/producers/crewai.py
"""CrewAI framework producer.

Wraps CrewAI ``step_callback`` / ``task_callback`` / ``Crew`` callback hooks
so each agent step, task completion, and crew run becomes an Action object.
Duck-typed against ``crewai`` — the producer does not import crewai, so the
SDK is not required at runtime.

Typical wiring::

    from crewai import Agent, Task, Crew
    from ancilis.producers import CrewAIActionProducer

    producer = CrewAIActionProducer(config=cfg, engine=engine)
    agent = Agent(role="researcher", step_callback=producer.step_callback("researcher"))
    task = Task(description="...", agent=agent, callback=producer.task_callback("research"))
    crew = Crew(agents=[agent], tasks=[task], step_callback=producer.step_callback("crew"))
    crew.kickoff()
"""

from __future__ import annotations

import hashlib
import json
import uuid
from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from ancilis.config import ResolvedConfig
from ancilis.engine.action import Action, ActionContext, ActionParameters, ToolInfo
from ancilis.engine.engine import Engine
from ancilis.engine.registry import ToolEntry, ToolRegistry, ToolStatus
from ancilis.engine.result import EvaluationResult
from ancilis.evidence.store import EvidenceStore
from ancilis.producers.protocol import ProducerType
from ancilis.telemetry import record_adapter_used

PROVIDER = "crewai"
PRODUCER_VERSION = "0.1.0"


@dataclass
class CrewAIEvent:
    """Normalized CrewAI hook event (step / task / crew)."""

    kind: str  # "step" | "task" | "crew"
    name: str
    agent_name: str
    output: Any = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class CrewAIObservation:
    action: Action
    evaluation: EvaluationResult


def _first_string_attr(obj: Any, attrs: tuple[str, ...]) -> str | None:
    for attr in attrs:
        value = getattr(obj, attr, None)
        if isinstance(value, str) and value:
            return value
    return None


def _first_string_key(mapping: Any, keys: tuple[str, ...]) -> str | None:
    if not isinstance(mapping, dict):
        return None
    for key in keys:
        value = mapping.get(key)
        if isinstance(value, str) and value:
            return value
    return None


def _step_name(step_output: Any, fallback: str) -> str:
    """Pull a useful identifier off a CrewAI step output object."""
    keys = ("tool", "tool_name", "agent_role", "agent", "name")
    return (
        _first_string_attr(step_output, keys)
        or _first_string_key(step_output, keys)
        or fallback
    )


def _task_name(task_output: Any, fallback: str) -> str:
    keys = ("description", "task_id", "name", "agent_role")
    return (
        _first_string_attr(task_output, keys)
        or _first_string_key(task_output, keys)
        or fallback
    )


def _crew_name(crew_output: Any, fallback: str) -> str:
    keys = ("name", "id", "crew_id")
    return (
        _first_string_attr(crew_output, keys)
        or _first_string_key(crew_output, keys)
        or fallback
    )


def _serializable(value: Any) -> Any:
    """Best-effort conversion for objects that may not JSON-encode."""
    if value is None or isinstance(value, (str, int, float, bool, list, dict)):
        return value
    for attr in ("model_dump", "dict", "to_dict"):
        method = getattr(value, attr, None)
        if callable(method):
            try:
                return method()
            except Exception:  # pragma: no cover - upstream object misbehaving
                continue
    return repr(value)


class CrewAIActionProducer:
    """Underlying producer that translates ``CrewAIEvent`` objects into Actions.

    Use the ``step_callback`` / ``task_callback`` / ``crew_callback`` factories
    to obtain closures with the signatures CrewAI expects.
    """

    def __init__(
        self,
        config: ResolvedConfig,
        engine: Engine,
        registry: ToolRegistry | None = None,
        evidence_store: EvidenceStore | None = None,
    ) -> None:
        self._config = config
        self._engine = engine
        self._registry = registry or engine.registry
        self._evidence_store = evidence_store if evidence_store is not None else EvidenceStore(config)
        self._session_id = str(uuid.uuid4())
        record_adapter_used(PROVIDER)

    @property
    def session_id(self) -> str:
        return self._session_id

    @property
    def producer_type(self) -> ProducerType:
        return ProducerType.FRAMEWORK

    @property
    def producer_version(self) -> str:
        return PRODUCER_VERSION

    def _tool_name(self, event: CrewAIEvent) -> str:
        return f"{PROVIDER}:{event.kind}:{event.name}"

    def translate(self, raw_invocation: CrewAIEvent) -> Action:
        payload = {
            "provider": PROVIDER,
            "kind": raw_invocation.kind,
            "name": raw_invocation.name,
            "output": _serializable(raw_invocation.output),
            "metadata": raw_invocation.metadata,
        }
        param_hash = hashlib.sha256(
            json.dumps(payload, sort_keys=True, default=repr).encode()
        ).hexdigest()
        tool_name = self._tool_name(raw_invocation)
        entry = self._registry.lookup(tool_name)
        dc_codes: list[str] = []
        for codes in self._config.data_classifications.values():
            for code in codes:
                if code not in dc_codes:
                    dc_codes.append(code)
        return Action(
            action_id=str(uuid.uuid4()),
            timestamp=datetime.now(timezone.utc).isoformat(),
            agent_id=raw_invocation.agent_name,
            source_type=self.producer_type.value,
            agent_owner=self._config.agent_owner or None,
            action_type="tool_call" if raw_invocation.kind == "step" else "api_request",
            tool=ToolInfo(
                name=tool_name,
                server=PROVIDER,
                description_hash=entry.description_hash if entry else None,
            ),
            parameters=ActionParameters(raw=payload, parameter_hash=param_hash),
            context=ActionContext(
                data_classifications=dc_codes,
                active_overlays=list(self._config.active_overlays.keys()),
                session_id=self._session_id,
            ),
            producer_type=self.producer_type.value,
            producer_version=self.producer_version,
        )

    def compute_tool_hash(self, tool_identifier: Any) -> str:
        return hashlib.sha256(str(tool_identifier).encode()).hexdigest()

    def register_tools(self, registry: ToolRegistry) -> list[str]:
        return [entry.name for entry in registry.get_all()]

    def _ensure_registered(self, event: CrewAIEvent) -> str:
        name = self._tool_name(event)
        if self._registry.lookup(name) is None:
            status = (
                ToolStatus.APPROVED
                if name in self._config.tools_allowed
                else ToolStatus.OBSERVED
            )
            self._registry.register(
                ToolEntry(
                    name=name,
                    description_hash=self.compute_tool_hash(name),
                    status=status,
                    approved_by="config" if status == ToolStatus.APPROVED else None,
                )
            )
        return name

    def observe(self, event: CrewAIEvent) -> CrewAIObservation:
        tool_name = self._ensure_registered(event)
        action = self.translate(event)
        evaluation = self._engine.evaluate(action)
        self._evidence_store.store(evaluation, tool_name=tool_name)
        return CrewAIObservation(action=action, evaluation=evaluation)

    # --- CrewAI-shaped callback factories ---

    def step_callback(self, agent_name: str | None = None) -> Callable[[Any], None]:
        """Return a function with the signature CrewAI expects for ``step_callback``.

        CrewAI calls ``step_callback(step_output)`` after each ReAct-style step.
        """

        agent = agent_name or self._config.agent_name

        def callback(step_output: Any) -> None:
            event = CrewAIEvent(
                kind="step",
                name=_step_name(step_output, "step"),
                agent_name=agent,
                output=step_output,
            )
            self.observe(event)

        return callback

    def task_callback(self, agent_name: str | None = None) -> Callable[[Any], None]:
        """Return a function for the per-Task ``callback=`` argument."""

        agent = agent_name or self._config.agent_name

        def callback(task_output: Any) -> None:
            event = CrewAIEvent(
                kind="task",
                name=_task_name(task_output, "task"),
                agent_name=agent,
                output=task_output,
            )
            self.observe(event)

        return callback

    def crew_callback(self, agent_name: str | None = None) -> Callable[[Any], None]:
        """Return a function suitable for crew-level step or completion callbacks."""

        agent = agent_name or self._config.agent_name

        def callback(crew_output: Any) -> None:
            event = CrewAIEvent(
                kind="crew",
                name=_crew_name(crew_output, "crew"),
                agent_name=agent,
                output=crew_output,
            )
            self.observe(event)

        return callback
